plugins/formal-verification/Uniqueness.md

```diff
@@ -0,0 +1,187 @@
+
+A unique variable is one whose value is null or else refers to an unshared object, one referred to by no other
+variables. This situation is called the _uniqueness invariant_. Parameters, receivers and return values may be declared
+unique. A use of a unique variable is a destructive read: the variable is atomically set to null at the same time that
+the value is read. The destruction preserves the _uniqueness invariant_.
+
+## Alias Burying and relaxation of uniqueness invariant
+
+Ensuring the _uniqueness invariant_ requires unique variables to be set null every time they are
+read. [Alias burying][2] keeps the uniqueness invariant but only require it to be true when it is needed. So the
+uniqueness invariant is thus not actually true at every point in the program, but the points when it is false are
+‘uninteresting’.
+
+## Annotations sets
+
+- All fields and return values must be annotated with either `Unique` or `Shared`.
+- Method parameters (including the receiver) must be annotated with one of these annotations:
+  `Unique`, `Borrowed`, `Shared`. \
+  Moreover, a parameter can be both `Unique` and `Shared`.
+- [Existing literature][4] shows that in systems similar to ours, annotations on
+  fields, return values and parameters are enough to infer annotation for local variable declarations.
+
+[//]: # (- It should be possible to infer annotations of variable declarations)
+
+## Annotations meaning
+
+- `Unique` denotes ownership, the value is only stored at this location.
+- A `Shared` annotation denotes that the variable can be accessed by outside objects, untracked aliases may exist.
+- `Borrowed` method parameter ensures that no further aliases are created by the method.
+
+## Smart casts and uniqueness
+
+Moving back to the [smart cast example](#smart-casts), we can see how our annotation system can help the compiler
+performing a smart cast.
+
+```kt
+class X(var n: Int?)
+
+fun useX(x: @Unique X): Int =
+    if (x.n != null) {
+        x.n // Uniqueness grants that 'x.n' have not changed after checking is nullability
+    } else {
+        0
+    }
+```
+
+Note that in this example, `useX` doesn't create new aliases of the parameter `x` and so it is possible to annotate it
+as `Borrowed`. Annotating parameter `x` as `Borrowed` preserves uniqueness after `useX` returns.
+
+```kt
+fun useX(x: @Borrowed @Unique X): Int {
+    // ...
+}
+
+fun main() {
+    val a = X(1) // Uniqueness here can be inferred
+    val y = useX(a)
+    // since the parameter 'x' in 'useX' is borrowed, 'a' is still unique after the call
+    val z = useX(a)
+}
+```
+
+## Checking annotations
+
+[Solving Shape-Analysis Problems in Languages with Destructive Updating][3] is the approach used by the authors
+of [Alias Burying][2] to statically check the correctness of uniqueness annotations. However, the formal verification
+plugin that we are developing is based on separation logic and should also be able to check the correctness of these
+annotations.
```

Comment:
Let's leave it at there existing checkers for now; the reference is good, but saying we'll use separation logic is probably not the direction we want to go in.